client/state/experiments/reducer.ts

```diff
@@ -0,0 +1,68 @@
+/**
+ * External Dependencies
+ */
+import { Action } from 'redux';
+
+/**
+ * Internal Dependencies
+ */
+import { EXPERIMENT_FETCH, EXPERIMENT_ASSIGN, CURRENT_USER_RECEIVE } from 'state/action-types';
+import { ExperimentState, ExperimentAssign } from 'state/experiments/types';
+import { tracksAnonymousUserId } from 'lib/analytics/ad-tracking';
+
+/**
+ * Attempt to get the anon id for the user, if set
+ */
+export function getAnonIdFromCookie(): string | null {
+	const id = tracksAnonymousUserId();
+
+	return id == null || id === '' ? null : id;
+}
+
+const appStartedAt = Date.now();
+
+const resetState: ( anonId: string | null ) => ExperimentState = anonId => ( {
+	anonId,
+	isLoading: true,
+	nextRefresh: appStartedAt,
+	variations: null,
+} );
+
+export default function reducer( state: ExperimentState = resetState( null ), action: Action ) {
+	switch ( action.type ) {
+		/**
+		 * Store the anon id when we first load the application
+		 */
+		case '@@INIT':
+			state.anonId = getAnonIdFromCookie();
+			break;
+
+		/**
+		 * Store the user's assignment from the API
+		 */
+		case EXPERIMENT_ASSIGN:
+			state = {
+				...state,
+				isLoading: false,
+				variations: ( action as ExperimentAssign ).variations,
+				nextRefresh: ( action as ExperimentAssign ).nextRefresh,
+			};
+			break;
+
+		/**
+		 * Start retrieving the user's assignment from the API
+		 */
+		case EXPERIMENT_FETCH:
+			state.isLoading = true;
+			break;
+
+		/**
+		 * When the user changes, we need to redetermine assigned variations
+		 */
+		case CURRENT_USER_RECEIVE:
+			state = resetState( state.anonId );
```

Comment:
The current user is ever set only once, at the very beginning of Calypso boot. It's guaranteed to never change after that, while Calypso is already running. If your reducer doesn't need to use some data from the `action.user` object to set its state, it doesn't need to react to the `CURRENT_USER_RECEIVE` action.